netmon/netmon.go

```diff
@@ -0,0 +1,613 @@
+// The reason for not importing them now, is because importing the whole agent
+// protocol adds up too much overhead because of the grpc protocol involved.
+
+// IPFamily define the IP address family type.
+type IPFamily int32
+
+// IPAddress describes the IP address format expected by Kata API.
+type IPAddress struct {
+	Family  IPFamily `json:"family,omitempty"`
+	Address string   `json:"address,omitempty"`
+	Mask    string   `json:"mask,omitempty"`
+}
+
+// Interface describes the network interface format expected by Kata API.
+type Interface struct {
+	Device      string       `json:"device,omitempty"`
+	Name        string       `json:"name,omitempty"`
+	IPAddresses []*IPAddress `json:"IPAddresses,omitempty"`
+	Mtu         uint64       `json:"mtu,omitempty"`
+	HwAddr      string       `json:"hwAddr,omitempty"`
+	PciAddr     string       `json:"pciAddr,omitempty"`
+}
+
+// Route describes the network route format expected by Kata API.
+type Route struct {
+	Dest    string `json:"dest,omitempty"`
+	Gateway string `json:"gateway,omitempty"`
+	Device  string `json:"device,omitempty"`
+	Source  string `json:"source,omitempty"`
+	Scope   uint32 `json:"scope,omitempty"`
+}
+
+const (
+	netmonName = "kata-netmon"
+
+	kataCmd              = "kata-network"
+	kataCLIAddIfaceCmd   = "add-iface"
+	kataCLIDelIfaceCmd   = "del-iface"
+	kataCLIUpdtRoutesCmd = "update-routes"
+
+	kataSuffix = "kata"
+
+	// For simplicity the code will only focus on IPv4 addresses for now.
+	netlinkFamily = netlink.FAMILY_V4
+
+	storageParentPath = "/var/run/kata-containers/netmon/sbs"
+	storageDirPerm    = 0750
+
+	// sharedFile is the name of the file that will be used to share
+	// the data between this process and the kata-runtime process
+	// responsible for updating the network.
+	sharedFile      = "shared.json"
+	storageFilePerm = 0640
+)
+
+// version is the netmon version. This variable is populated at build time.
+var version = "unknown"
+
+type netmonParams struct {
+	sandboxID   string
+	runtimePath string
+	debug       bool
+	logLevel    string
+}
+
+type netmon struct {
+	netmonParams
+
+	storagePath string
+	sharedFile  string
+
+	netIfaces map[int]Interface
+
+	linkUpdateCh chan netlink.LinkUpdate
+	linkDoneCh   chan struct{}
+
+	rtUpdateCh chan netlink.RouteUpdate
+	rtDoneCh   chan struct{}
+
+	netHandler *netlink.Handle
+}
+
+var netmonLog = logrus.New()
+
+func printVersion() {
+	fmt.Printf("%s version %s\n", netmonName, version)
+}
+
+func parseOptions() netmonParams {
+	var version bool
+
+	params := netmonParams{}
+
+	flag.BoolVar(&params.debug, "d", false, "enable debug mode")
+	flag.BoolVar(&version, "v", false, "display program version and exit")
+	flag.StringVar(&params.sandboxID, "s", "", "sandbox id (required)")
+	flag.StringVar(&params.runtimePath, "r", "", "runtime path (required)")
+	flag.StringVar(&params.logLevel, "log", "warn",
+		"log messages above specified level: debug, warn, error, fatal or panic")
+
+	flag.Parse()
```

Comment:
Not sure if it's possible with the `flag` package (?), but it would be useful if running `./kata-netmon -h` could show a sentence or two on what this new component actually does. If flag doesn't allow that, how about switching to https://github.com/urfave/cli (or https://github.com/spf13/cobra :)?

Reply:
I would like to keep this component simple, and pulling down `urfave/cli` will make it bigger. I'll see if what you ask for is possible with `flag` but I don't want to introduce a heavy package to handle the very simple command line.

Reply:
Yeah, I understand. Although it might not be possible for `kata-netmon -h` to show this, we *could* make `kata-netmon help` display the human-readable text, then call `flag.PrintDefaults()`?

Reply:
I have been able to override `-h`